src/calliope/config.py

```diff
@@ -0,0 +1,369 @@
+    """
+    Whether or not to save the result of each SPORES mode run between iterations.
+    If False, will consolidate all iterations into one dataset after completion of N iterations (defined by `number`) and save that one dataset.
+    """
+
+    save_per_spore_path: Path | None = Field(default=None)
+    """If saving per spore, the path to save to."""
+
+    skip_cost_op: bool = Field(default=False)
+    """If the model already contains `plan` mode results, use those as the initial base run results and start with SPORES iterations immediately."""
+
+    @model_validator(mode="after")
+    def require_save_per_spore_path(self) -> Self:
+        """Ensure that path is given if saving per spore."""
+        if self.save_per_spore:
+            if self.save_per_spore_path is None:
+                raise ValueError(
+                    "Must define `save_per_spore_path` if you want to save each SPORES result separately."
+                )
+            elif not self.save_per_spore_path.is_dir():
+                raise ValueError("`save_per_spore_path` must be a directory.")
+        return self
+
+
+class Solve(ModeBaseModel):
+    """Base configuration options used when solving a Calliope optimisation problem (`calliope.Model.solve`)."""
+
+    model_config = {
+        "title": "solve",
+        "extra": "forbid",
+        "revalidate_instances": "always",
+        "json_schema_extra": hide_from_schema(["mode"]),
+    }
+
+    save_logs: Path | None = Field(default=None)
+    """If given, should be a path to a directory in which to save optimisation logs."""
+
+    solver_io: str | None = Field(default=None)
+    """
+    Some solvers have different interfaces that perform differently.
+    For instance, setting `solver_io="python"` when using the solver `gurobi` tends to reduce the time to send the optimisation problem to the solver.
+    """
+
+    solver_options: dict = Field(default={})
+    """Any solver options, as key-value pairs, to pass to the chosen solver"""
+
+    solver: str = Field(default="cbc")
+    """Solver to use. Any solvers that have Pyomo interfaces can be used. Refer to the Pyomo documentation for the latest list."""
+
+    zero_threshold: float = Field(default=1e-10)
+    """On postprocessing the optimisation results, values smaller than this threshold will be considered as optimisation artefacts and will be set to zero."""
+
+    shadow_prices: UniqueList[str] = Field(default=[])
+    """Names of model constraints."""
+
+    spores: SolveSpores = SolveSpores()
+
+
+class CalliopeConfig(ConfigBaseModel):
+    """Calliope configuration class."""
+
+    model_config = {"title": "config"}
+    init: Init = Init()
+    build: Build = Build()
+    solve: Solve = Solve()
+
+    @model_validator(mode="before")
+    @classmethod
+    def update_solve_mode(cls, data):
+        """Solve mode should match build mode."""
+        data["solve"]["mode"] = data["build"]["mode"]
+        return data
```

Comment:
if you do `model.config.update({"build": {"mode": "operate"}})` it will update the solve "mode" too. The solve "mode" is really just a tracker of the build mode, and this ensures they are always in sync.

Reply:
I meant it as 'this is not used anywhere in our code'. But it seems like its an update to `pydantic` behaviour (fancy!).

Having `mode` in two different places is an issue, though. Even if `pydantic` ensures they are equal, it introduces unnecessary ambiguity. 

`Model.solve()` seems to update `solve.mode`, and then immediately use it, and then it's used nowhere else after.

```python

        kwargs["mode"] = self.config.build.applied_keyword_overrides.get(
            "mode", self.config.build.mode
        )

        this_solve_config = self.config.update({"solve": kwargs}).solve
        self._model_data.attrs["timestamp_solve_start"] = log_time(
            LOGGER,
            self._timings,
            "solve_start",
            comment=f"Optimisation model | starting model in {this_solve_config.mode} mode.",
        )

        shadow_prices = this_solve_config.shadow_prices
        self.backend.shadow_prices.track_constraints(shadow_prices)

        if this_solve_config.mode == "operate":
            results = self._solve_operate(**this_solve_config.model_dump())
        else:
            results = self.backend._solve(
                warmstart=warmstart, **this_solve_config.model_dump()
            )
```

If we just used the value `build.mode` within `Model.solve()`, all this code is unnecessary.
Perhaps ensuring that `Model._is_built` is automatically set to `False` if the mode is ever updated is a better approach?